macos-redirector/network-extension/PacketTunnelProvider.swift

```diff
@@ -0,0 +1,102 @@
+        let ipv4Settings = NEIPv4Settings(addresses: ["10.0.0.1"], subnetMasks: ["255.255.255.0"])
+        ipv4Settings.includedRoutes = [
+            NEIPv4Route.default()
+        ]
+        tunnelNetworkSettings.ipv4Settings = ipv4Settings
+        tunnelNetworkSettings.dnsSettings = NEDNSSettings(servers: ["10.0.0.53"])
+        
+        let addr = self.protocolConfiguration.serverAddress!
+                
+        let proxy_file: FileHandle, redir_file: FileHandle
+        do {
+            proxy_file = try FileHandle(forReadingFrom: URL(fileURLWithPath: "\(addr).proxy"));
+            redir_file = try FileHandle(forWritingTo: URL(fileURLWithPath: "\(addr).redir"))
+        } catch {
+            return completionHandler(error)
+        }
+        
+        self.setTunnelNetworkSettings(tunnelNetworkSettings) { err in
+            completionHandler(err)
+            log.debug("tunnel settings set (err=\(err))")
+            if err == nil {
+                DispatchQueue.global().async {
+                    self.redirectPackets(redir_file)
+                }
+                DispatchQueue.global().async {
+                    self.reinjectPackets(proxy_file)
+                }
+            }
+        }
+        
+        log.debug("startTunnel done")
+    }
+    
+    func redirectPackets(_ wfile: FileHandle) {
+        log.debug("redirecting packets to \(wfile, privacy: .public)...")
+        self.packetFlow.readPacketObjects { packets in
+            for packet in packets {
+                
+                var message = Mitmproxy_Ipc_FromRedirector()
+                message.packet.data = packet.data
+                message.packet.pid = 0 // TODO
+                message.packet.processName = packet.metadata!.sourceAppSigningIdentifier
+                
+                do {
+                    try writeIpcMessage(message: message, fh: wfile)
+                } catch {
+                    log.error("redirectPackets errored: \(error)")
+                    exit(1)
+                }
+                self.redirectPackets(wfile)
+            }
+        }
+    }
+    
+    func reinjectPackets(_ rfile: FileHandle) {
+        log.debug("reading \(rfile, privacy: .public)...")
+        do {
+            while let packet = try readIpcMessage(ofType: Mitmproxy_Ipc_Packet.self, fh: rfile) {
+                
+                log.debug("reinjecting packet...")
+                self.packetFlow.writePackets([packet.data], withProtocols: [AF_INET as NSNumber])
+                
+            }
+        } catch {
+            log.error("redirectPackets errored: \(error)")
+            exit(1)
+        }
+        
+        log.debug("exiting...")
+        exit(0)
+    }
+    
+    override func stopTunnel(with reason: NEProviderStopReason, completionHandler: @escaping () -> Void) {
+        log.debug("stopTunnel \(String(describing:reason))")
+        completionHandler()
+        exit(0) // XXX: is that a reasonable approach to tear things down?
```

Comment:
Or maybe create a function to handle "panic", that clean VPNmanager before exiting?

Reply:
and maybe also it could send to pipe a message to rust, so we can remove also macos-redirector from /applications

Reply:
That's not possible AFAIK, the network extension cannot manage VPN configurations created by the app (it does not seem them at all).

Reply:
you are right, I think we could send a termination signal to mitmproxy that report it to the app again, but probably not here